### PR TITLE
fix: Upgrade to Python 3.12

### PR DESCRIPTION
**Pull request by faucomte97**
This change is [Reviewable](https://reviewable.io/reviews/ocadotechnology/rapid-router/1713)

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -6,24 +6,26 @@ on:
...
-          python-version: "3.8"
+          python-version: "3.12"
           architecture: "x64"
       - name: Install pip requirements
         run: |
-          pip install pipenv==2022.10.4
+          pip install pipenv
           pip install pytest-cov
           pipenv install --dev --system
+          pip install -U "git+https://github.com/ocadotechnology/codeforlife-portal@python312#egg=cfl-common&subdirectory=cfl_common"
```
Comment (SKairinos) [inline .github/workflows/ci.yml:27]: assuming this is temp. remember to remove.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -6,24 +6,26 @@ on:
...
+          pip install -U "git+https://github.com/ocadotechnology/codeforlife-portal@python312#egg=codeforlife-portal"
```
Comment (SKairinos) [inline .github/workflows/ci.yml:28]: assuming this is temp. remember to remove.

**Review by SKairinos — APPROVED** [review]

File: .github/workflows/snyk.yaml [hunk 3]
```diff
@@ -26,7 +26,7 @@ jobs:
       - name: Generate requirements
         run: pip freeze > requirements.txt
       - name: Run Snyk to check for vulnerabilities in Python
-        uses: snyk/actions/python-3.8@master
+        uses: snyk/actions/python-3.10@master
```
Comment (SKairinos) [inline .github/workflows/snyk.yaml:29]: why have you hardcoded this to 3.10? why not 3.12?

**Review by SKairinos — CHANGES_REQUESTED** [review]

**Review by faucomte97 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/ocadotechnology/rapid-router/1713)* status: 0 of 7 files reviewed, 3 unresolved discussions (waiting on @faucomte97 and @SKairinos)

___
*[`.github/workflows/ci.yml` line 27 at r1](https://reviewable.io/reviews/ocadotechnology/rapid-router/1713#gh-1740782579:-O5myua49JEGVEuY4BX7:b-896fix) ([raw file](https://github.com/ocadotechnology/rapid-router/blob/9ad20c8e486bd8b62f4ed383af9b3cd55d362c7f/.github/workflows/ci.yml#L27)):*
<details><summary><i>Previously, SKairinos (Stefan Kairinos) wrote…</i></summary><blockquote>

assuming this is temp. remember to remove.
</blockquote></details>

Done.
___
*[`.github/workflows/ci.yml` line 28 at r1](https://reviewable.io/reviews/ocadotechnology/rapid-router/1713#gh-1740782999:-O5myuvv8CiGpkKt2ABQ:b-896fix) ([raw file](https://github.com/ocadotechnology/rapid-router/blob/9ad20c8e486bd8b62f4ed383af9b3cd55d362c7f/.github/workflows/ci.yml#L28)):*
<details><summary><i>Previously, SKairinos (Stefan Kairinos) wrote…</i></summary><blockquote>

assuming this is temp. remember to remove.
</blockquote></details>

Done.
___
*[`.github/workflows/snyk.yaml` line 29 at r1](https://reviewable.io/reviews/ocadotechnology/rapid-router/1713#gh-1740793537:-O5myvbw1P7GHOM4YbiT:b-896fix) ([raw file](https://github.com/ocadotechnology/rapid-router/blob/9ad20c8e486bd8b62f4ed383af9b3cd55d362c7f/.github/workflows/snyk.yaml#L29)):*
<details><summary><i>Previously, SKairinos (Stefan Kairinos) wrote…</i></summary><blockquote>

why have you hardcoded this to 3.10? why not 3.12?
</blockquote></details>

Done.

**Review by SKairinos — APPROVED** [review]
> Reviewed 5 of 7 files at r1, 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/ocadotechnology/rapid-router/1713)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @faucomte97)

**Pull request merged**